precompiles/balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+		// Parse input.
+		input.expect_arguments(0)?;
+
+		// Fetch info.
+		let amount: U256 = pallet_balances::Pallet::<Runtime, Instance>::total_issuance().into();
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(amount).build(),
+			logs: vec![],
+		})
+	}
+
+	fn balance_of(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+		// Read input.
+		input.expect_arguments(1)?;
+
+		let owner: H160 = input.read::<Address>()?.into();
+
+		// Fetch info.
+		let amount: U256 = {
+			let owner: Runtime::AccountId = Runtime::AddressMapping::into_account_id(owner);
+			pallet_balances::Pallet::<Runtime, Instance>::usable_balance(&owner).into()
+		};
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(amount).build(),
+			logs: vec![],
+		})
+	}
+
+	fn allowance(mut input: EvmDataReader, target_gas: Option<u64>) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+		// Read input.
+		input.expect_arguments(2)?;
+
+		let owner: H160 = input.read::<Address>()?.into();
+		let spender: H160 = input.read::<Address>()?.into();
+
+		// Fetch info.
+		let amount: U256 = {
+			let owner: Runtime::AccountId = Runtime::AddressMapping::into_account_id(owner);
+			let spender: Runtime::AccountId = Runtime::AddressMapping::into_account_id(spender);
+
+			ApprovesStorage::<Runtime, Instance>::get(owner, spender)
+				.unwrap_or_default()
+				.into()
+		};
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(amount).build(),
+			logs: vec![],
+		})
+	}
+
+	fn approve(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let spender: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		// Write into storage.
+		{
+			let caller: Runtime::AccountId =
+				Runtime::AddressMapping::into_account_id(context.caller);
+			let spender: Runtime::AccountId = Runtime::AddressMapping::into_account_id(spender);
+			let amount = Self::u256_to_amount(amount)?;
+
+			ApprovesStorage::<Runtime, Instance>::insert(caller, spender, amount);
```

Comment:
Replaced. However users must be carefull with it since it's vulnerable to front-running :
https://blog.smartdec.net/erc20-approve-issue-in-simple-words-a41aaf47bca6

However that's how the ERC20 standard is designed.

Reply:
It would be nice to add a comment that links the ERC20 spec. And that specifies that the expected behavior for `approve` is indeed the replacement of the amount.

Reply:
> "Some ERC20 token contracts implement two more functions: increaseApproval and decreaseAppoval. These functions only change the allowance by a certain value, instead of setting the new one. In the example above Alice could have called the decreaseApproval function with the value equal to 700. As a result, Bob could transfer only 1000 tokens at maximum, instead of 1300. Personally, I recommend using this option if possible."

So, it would be nice to add these two functions.

Reply:
We can add that in a follow up PR.